### ISSUE TITLE
Loss curve changes after checkpoint load

### ISSUE THREAD

**Issue by okoge-kaz** [microsoft/Megatron-DeepSpeed#274]
## Overview

I am training 175B model using Megatron-DeepSpeed. 
The loss changes when the checkpoint is loaded and training is resumed. 

### Training script

175B: https://github.com/llm-jp/Megatron-DeepSpeed/blob/feature/fujii/abci/scripts/abci/175B/49node/dp2_tp4_pp49_zero1_flash_attn_rottary_bf16_100K.sh
175B ds_config: https://github.com/llm-jp/Megatron-DeepSpeed/blob/feature/fujii/abci/examples_deepspeed/rebase/ds_config_gpt_TEMPLATE_bf16.json
deepspeed version:  deepspeed               0.11.1

1.3B: https://github.com/llm-jp/Megatron-DeepSpeed/blob/feature/fujii/abci/scripts/abci/1.3B/dp16_tp1_pp1_zero1.sh
1.3B ds_config: https://github.com/llm-jp/Megatron-DeepSpeed/blob/feature/fujii/abci/examples_deepspeed/rebase/ds_config_gpt_TEMPLATE_bf16.json
deepspeed version:  deepspeed               0.11.1

### Cluster Info

175B: A100(40GB) 8GPU x 49 Node
1.3B: A100(40GB) 8GPU x 2 node

### 1.3B 
wandb project: https://wandb.ai/okoge/megatron-deepspeed-loss-spike/workspace

experiments: 
- https://wandb.ai/okoge/megatron-deepspeed-loss-spike/runs/h6axmsf0
- https://wandb.ai/okoge/megatron-deepspeed-loss-spike/runs/1o0y8dw6

At first glance, it looks fine.
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/5d43b74f-e76e-412f-87ad-7ac530efd819)

but when enlarged, the loss is misaligned.
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/52ac561f-7af7-4b37-8ee6-1af3b7821c28)
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/7b7196df-6e28-44e2-9a31-693d7270ec56)

### 175B 

this experiment's wandb log is private, sorry.
But the situation is as shown in the screenshots below

![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/a7a11f61-9dcf-4e9a-b27a-bcb8d8cb5c5e)

![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/3d4a1829-d68c-40e7-a328-fca820f3df06)

## Supplement Info

In the 1.3B training, I confirmed that using Rotary Positional Embeddings and Flash-Attention caused a change in the Loss Curve.

### No RoPE, Flash Attention
wandb runs:
- https://wandb.ai/okoge/megatron-deepspeed-loss-spike/runs/ja75vnil
- https://wandb.ai/okoge/megatron-deepspeed-loss-spike/runs/nh3bxmwi

![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/11702828-1c8b-4eb5-bb16-65cc19b0f839)
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/e12e953e-3a65-4ee4-9e86-52017e2c9ebc)

### Using RoPE, Flash Attention

wandb runs
- https://wandb.ai/okoge/megatron-deepspeed-loss-spike/runs/1o0y8dw6
- https://wandb.ai/okoge/megatron-deepspeed-loss-spike/runs/h6axmsf0

![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/fdec6166-f2fc-4444-8fbd-363a18a8eff3)

**Comment by okoge-kaz** [comment]
Related Issue: https://github.com/microsoft/Megatron-DeepSpeed/issues/248

**Comment by okoge-kaz** [comment]
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/24912932-04db-4b93-9ff9-fdf5e01ef646)

deepspeed version: 0.11.1

**Comment by tohtana** [comment]
@okoge-kaz Thank you for the detailed report. We reproduced the mismatch of loss values after loading a checkpoint. After running small-scale experiments, we found that some kernels do not reproduce the exact same results. Can you try the followings to make sure if you are seeing the same issue? I observed the loss values exactly matched after these changes.

- Add `--no-bias-dropout-fusion` and `--no-bias-gelu-fusion`
- Remove `--use-flash-attn`

**Comment by tohtana** [comment]
@okoge-kaz Just for clarification, can you let us know the version of Megatron-DeepSpeed you used?
I didn't find `--pp-partition-method` used in the [training script](https://github.com/llm-jp/Megatron-DeepSpeed/blob/feature/fujii/abci/scripts/abci/175B/49node/dp2_tp4_pp49_zero1_flash_attn_rottary_bf16_100K.sh) in the current version.

**Comment by okoge-kaz** [comment]
> @okoge-kaz Just for clarification, can you let us know the version of Megatron-DeepSpeed you used? I didn't find `--pp-partition-method` used in the [training script](https://github.com/llm-jp/Megatron-DeepSpeed/blob/feature/fujii/abci/scripts/abci/175B/49node/dp2_tp4_pp49_zero1_flash_attn_rottary_bf16_100K.sh) in the current version.


In our project, the vocab size is large(about 100K), so we have added a process that considers the embedding layer and lm-head as a transformer block, based on the implementation of the BLOOM project.

This is the implementation in BLOOM
https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/main/megatron/model/gpt_model.py#L310-L319

**Comment by okoge-kaz** [comment]
> @okoge-kaz Thank you for the detailed report. We reproduced the mismatch of loss values after loading a checkpoint. After running small-scale experiments, we found that some kernels do not reproduce the exact same results. Can you try the followings to make sure if you are seeing the same issue? I observed the loss values exactly matched after these changes.
> 
> * Add `--no-bias-dropout-fusion` and `--no-bias-gelu-fusion`
> * Remove `--use-flash-attn`

@tohtana
Thank you very much for your reply! 
I will try it as soon as I have some computing resources available.

By the way, Is it possible to avoid the checkpoint issue when using flash-attention?

**Comment by okoge-kaz** [comment]
As additional information, I also faced the phenomenon of loss curve change when checkpoints are loaded when I am using Megatron-LM for continuous learning of Llama-2. I hope this information will be useful for debugging.

![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/9e0e4b4b-4d86-4d56-bb19-d549c861c176)

**Comment by okoge-kaz** [comment]
@tohtana
Sorry for the late reply.
Adding ` --no-bias-dropout-fusion` and `--no-bias-gelu-fusion`and removing `--use-flash-attn`,  the loss matches exactly!

![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/12a807eb-999d-40c3-ac7e-7d1343e99c4b)
![Screenshot 2023-11-10 at 0 18 12](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/e1b16469-f23f-46d5-89c4-2da24c085f7d)

Thank you very much! 
By the way, is there any way to use flash-attention and RoPE at the same time and still the Loss match exactly?

This is a very important issue for us as we would like to use Megatron-DeepSpeed when we create our next model.

Thanks.

**Comment by tohtana** [comment]
@okoge-kaz Let me share some related information.

- The backward pass of FlashAttention is [non-deterministic](https://github.com/Dao-AILab/flash-attention/issues/349), meaning we cannot reproduce exact same results as long as we use FlashAttention.
- According to [README](https://github.com/Dao-AILab/flash-attention/#tests), FlashAttention produces larger numerical errors than the reference implementation of PyTorch. This could potentially impact training stability, especially in very large models.

Unfortunately, there's no definitive solution to these issues yet. Then, I do have a couple of suggestions:

- If you've only experimented with FlashAttention V1, consider trying out V2 or the Triton version. This [document](https://github.com/microsoft/Megatron-DeepSpeed/tree/main/examples_deepspeed/sequence_parallel) provides guidance on using the Triton version with Megatron-DeepSpeed.
- The primary cause of training instability is still unclear. Given that the training instability was observed after loading a checkpoint, I recommend adding `--no-bias-dropout-fusion` with FlashAttention enabled. This will show a non-deterministic behavior, but the stability might change if the issue is related to fused dropout.

**Comment by okoge-kaz** [comment]
@tohtana Thank you for your detailed investigation of the loss instability issue.

We will have the opportunity to train the 175B model in early December (but only for a few hours), so I will use the `--no-bias-dropout-fusion` argument you showed to check the instability of Loss.

Also, I was using `--pp-partition-method`, which was not in the original argument options of Megatron-DeepSpeed, so I will investigate the behavior without it.

I will keep the Issue Open to share findings here. 

Again, thanks for your detailed investigation.

**Comment by okoge-kaz** [comment]
Let me share the results of experiments.

1.3B DP=2, PP=2, TP=2 ZeRO Stage1 (A100 40GB x 8)
Flash Attention version 2.3.0, DeepSpeed version 0.11.1

1. **without** --pp-partition-method, **with** `--no-bias-dropout-fusion`
   script: https://github.com/llm-jp/Megatron-DeepSpeed/commit/d6648e99fc875ccc0092eb2cfc458215168f5d49#diff-511554af37bf0b364e3a9cb65bac8e8ea4499385c3827510a29b531c74111fba

   
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/15e0c731-126b-4ec7-8ec2-370ef6af4e9c)
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/9096d0be-27a4-414b-9892-59229b48136c)

Observing about 100 iterations after checkpoint load, the loss after checkpoint load is always high. Therefore, it seems that non-deteministic reasons are not the only reasons.

2. **without** --pp-partition-method, **without** `--no-bias-dropout-fusion`
     script: https://github.com/llm-jp/Megatron-DeepSpeed/commit/d6648e99fc875ccc0092eb2cfc458215168f5d49#diff-e61698a0b9307e2d7c1db383d4720d4d9b1bdc2d7cefb1bc7ed7ec4617edb53b

![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/7b147b64-15de-4554-af7c-fa5a49caed0a)
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/f2a1e6cc-6515-4579-b600-529f9c41fd42)

same above

I will also do some experiments on what happens when the DeepSpeed version is brought up to date.
And I am using also flash-attention version 1.x  to observe what happens.

(all logging data: https://wandb.ai/okoge/megatron-deepspeed-loss-spike/workspace?workspace=user-okoge)

**Comment by okoge-kaz** [comment]
1.3B DP=2, PP=2, TP=2 ZeRO Stage1 (A100 40GB x 8)
Flash Attention version 2.3.3, DeepSpeed version 0.12.3
PyTorch 2.1.0+cu118

1. **without** `--pp-partition-method`, with `--no-bias-dropout-fusion`
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/e9707d34-5e68-4ac3-a9cd-40848823307f)
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/9ba6b6b4-6d29-40d5-9b24-ad7e348435e4)
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/9adcd562-cdb4-43a9-9fd7-8f58056cc114)

wandb run1: https://wandb.ai/okoge/megatron-deepspeed-loss-spike/runs/mn07xosl
wandb run2: https://wandb.ai/okoge/megatron-deepspeed-loss-spike/runs/wgmtemzr
wandb run3: https://wandb.ai/okoge/megatron-deepspeed-loss-spike/runs/ngs5uu7j
     
2.  **without** `--pp-partition-method`, without `--no-bias-dropout-fusion`
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/df4dd3b3-e3af-4d99-a084-5c2f04b1a692)
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/d3b02cfe-f220-4048-9893-bc8da304ca01)
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/049c9579-b1ba-4a5d-865b-00367947de32)

wandb run1: https://wandb.ai/okoge/megatron-deepspeed-loss-spike/runs/k6z91rag
wandb run2: https://wandb.ai/okoge/megatron-deepspeed-loss-spike/runs/jo52aqd7
wandb run3: https://wandb.ai/okoge/megatron-deepspeed-loss-spike/runs/4zf8454c

I have changed the version of DeepSpeed to 0.12.3, but the loss is still high after checkpoint load.

**Comment by okoge-kaz** [comment]
1.3B DP=2, PP=2, TP=2 ZeRO Stage1 (A100 40GB x 8)
Flash Attention version 2.3.3, DeepSpeed version 0.12.3
PyTorch 2.1.0+cu118

1. with `--no-bias-dropout-fusion` and with `--no-bias-gelu-fusion`
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/f2918ea5-a68f-49a7-a513-6f21f8b0defb)
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/79e28598-5300-4bc4-94b4-09a27ca89a06)

wandb run1: https://wandb.ai/okoge/megatron-deepspeed-loss-spike/runs/g9whmwxb
wandb run2: https://wandb.ai/okoge/megatron-deepspeed-loss-spike/runs/1rzvz4s4

**Comment by okoge-kaz** [comment]
1.3B DP=2, PP=2, TP=2 ZeRO Stage1 (A100 40GB x 8)
Flash Attention version 1.0.4, DeepSpeed version 0.12.3
PyTorch 2.1.0+cu118

with `--no-bias-dropout-fusion` and with `--no-bias-gelu-fusion`

![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/358c7887-a697-4bd9-80d1-fa9d396240b8)
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/c40e9172-bad0-46de-b587-530aa2dec938)

I tried changing the version of FlashAttention, but that did not solve the issue.

In Megatron-LM, this phenomenon does not occur (only smaller loss curve gaps) when flash-attention is enabled, but why?

**Comment by tohtana** [comment]
There was a bug related to setting the dropout probability for FlashAttention.
It has been addressed in [This PR](#295).

**Comment by okoge-kaz** [comment]
Loss no longer significantly shifted after checkpoint load even with flash-attn enabled.
Thank you very much! Very helpful!



![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/6b551e51-ef72-4b11-8961-4a80a1342e26)
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/7463ea57-eff9-43b6-908e-f8252c902b69)

**Comment by okoge-kaz** [comment]
Thanks, @tohtana for investigating the checkpoint load bug very precisely.
